src/Rector/FuncCall/TypeHintTappableCallRector.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+namespace RectorLaravel\Rector\FuncCall;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\Closure;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Param;
+use PHPStan\Type\ObjectType;
+use Rector\NodeTypeResolver\TypeComparator\TypeComparator;
+use Rector\PHPStanStaticTypeMapper\Enum\TypeKind;
+use Rector\Rector\AbstractRector;
+use Rector\StaticTypeMapper\StaticTypeMapper;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \RectorLaravel\Tests\Rector\FuncCall\TypeHintTappableCallRector\TypeHintTappableCallRectorTest
+ */
+class TypeHintTappableCallRector extends AbstractRector
+{
+    private const TAPPABLE_TRAIT = 'Illuminate\Support\Traits\Tappable';
+
+    public function __construct(
+        private readonly TypeComparator $typeComparator,
+        private readonly StaticTypeMapper $staticTypeMapper
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Automatically type hints your tappable closures',
+            [
+                new CodeSample(<<<'CODE_SAMPLE'
+tap($collection, function ($collection) {}
+CODE_SAMPLE,
+                    <<<'CODE_SAMPLE'
+tap($collection, function (Collection $collection) {}
+CODE_SAMPLE
+                ),
+                new CodeSample(<<<'CODE_SAMPLE'
+(new Collection)->tap(function ($collection) {}
+CODE_SAMPLE,
+                    <<<'CODE_SAMPLE'
+(new Collection)->tap(function (Collection $collection) {}
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    public function getNodeTypes(): array
+    {
+        return [MethodCall::class, FuncCall::class];
+    }
+
+    /**
+     * @param  MethodCall|FuncCall  $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (! $this->isName($node->name, 'tap')) {
+            return null;
+        }
+
+        if ($node->isFirstClassCallable()) {
+            return null;
+        }
+
+        if ($node instanceof MethodCall && $node->getArgs() !== []) {
+            return $this->refactorMethodCall($node);
+        }
+
+        if (count($node->getArgs()) < 2 || ! $node->getArgs()[1]->value instanceof Closure) {
+            return null;
+        }
+
+        /** @var Closure $closure */
+        $closure = $node->getArgs()[1]->value;
+
+        $this->refactorParameter($closure->getParams()[0], $node->getArgs()[0]->value);
```

Comment:
I think we need an extra check here for the number of params. It's unlikely, but the rule would fail if it encounters a Closure without params: `tap('test', function () {});`.